go/vt/vtgate/planbuilder/operators/rewrite/rewriters.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rewrite
+
+import (
+	"vitess.io/vitess/go/vt/vtgate/planbuilder/operators/ops"
+)
+
+type (
+	Func          func(ops.Operator) (ops.Operator, TreeIdentity, error)
+	BreakableFunc func(ops.Operator) (ops.Operator, TreeIdentity, VisitRule, error)
+
+	// TreeIdentity tracks modifications to node and expression trees.
+	// Only return SameTree when it is acceptable to return the original
+	// input and discard the returned result as a performance improvement.
+	TreeIdentity bool
+
+	// VisitRule signals to the rewriter if the children of this operator should be visited or not
+	VisitRule bool
+)
+
+const (
+	SameTree TreeIdentity = false
+	NewTree  TreeIdentity = true
+
+	VisitChildren VisitRule = true
+	SkipChildren  VisitRule = false
+)
+
+// Visit allows for the walking of the operator tree. If any error is returned, the walk is aborted
+func Visit(root ops.Operator, visitor func(ops.Operator) error) error {
+	_, err := TopDown(root, func(op ops.Operator) (ops.Operator, TreeIdentity, VisitRule, error) {
+		err := visitor(op)
+		if err != nil {
+			return nil, SameTree, SkipChildren, err
+		}
+		return op, SameTree, VisitChildren, nil
+	})
+	return err
+}
+
+// BottomUp rewrites an operator tree from the bottom up. BottomUp applies a transformation function to
+// the given operator tree from the bottom up. Each callback [f] returns a TreeIdentity that is aggregated
+// into a final output indicating whether the operator tree was changed.
+func BottomUp(root ops.Operator, f Func) (ops.Operator, error) {
+	op, _, err := bottomUp(root, f)
```

Comment:
looks like the comment should actually move to `bottomUp` function where the logic resides.

Reply:
nah, the comment belongs here. the internal `bottomUp` does the actual work, but the description belongs here, I think.